Add pyrate-limiter and apply CV suggested rate limit

mizaki opened the pull request:
This add pyrate-limiter to the required deps.

I took the (very) old comment from the CV API [forum](https://comicvine.gamespot.com/forums/api-developers-2334/api-rate-limiting-1746419/) and set the rate limit to 1 every 2 seconds. I've left a simple check and 10 second wait as a fallback. Instead of waiting many multiple minutes which might make the (GUI) user think the program has hung, it provides a network error with a link where to check your API usage.

mizaki committed:
Add pyrate-limiter and apply CV suggested rate limit

mizaki commented:
Build error is: E: Failed to fetch http://azure.archive.ubuntu.com/ubuntu/pool/main/m/mesa/libegl-mesa0_22.2.5-0ubuntu0.1%7e22.04.1_amd64.deb  404  Not Found [IP: 40.119.46.219 80]

mizaki committed:
Remove wait on CV rate limit in autotag

lordwelch commented:
So with the current limit of (1, 2) this means that there is at least 2 seconds between every request which more than doubles the time it takes to auto-tag ~50 comics that I tested with. At a minimum this should be (1, 1) but I would rather go with (10, 10). I also want a separate more severe limiter for the shared api key as it can run up api requests faster.

mizaki commented:
> So with the current limit of (1, 2) this means that there is at least 2 seconds between every request which more than doubles the time it takes to auto-tag ~50 comics that I tested with. At a minimum this should be (1, 1) but I would rather go with (10, 10). I also want a separate more severe limiter for the shared api key as it can run up api requests faster.

I was going by
```
"Space out your requests so AT LEAST one second passes between each and you can make requests all day. Go even a millisecond faster and you'll hit a brick wall REALLY HARD."
```
which doesn't appear to be true. Which I interpreted to mean a gap of 1 second but 1,1 is probably correct. Of course, getting any answer from CV seems very unlikely. Other than the status message, I can't seem to find if the API reports the cool down required (like a HTTP header), going to the website appears to be the only resort...

It also appears it uses the correct [HTTP code](https://comicvine.gamespot.com/forums/api-developers-2334/new-api-http-status-when-rate-limited-1751775/) now for rate limit so I'll switch over to that. 107 is no longer listed as a status code.

It is possible to stack the limit for different [durations](https://github.com/vutran1710/PyrateLimiter#basic-usage). I'm not sure if that's of any benefit here in say a limit per minute and hour instead of seconds.

A option could be added (with a message about the default key being hard coded)? Have a spin box so it can be limited.

mizaki committed:
Move rate limit check from defunc API status code 107 to HTTP code 429. Set a limit of 10 request every 10 seconds except for the default API key which is 1,2 (to be finisalised). Remove wait on rate limit option.

mizaki commented:
I've changed to default to 10,10 and added 1,2 for the default key. What number should the default key be, 1,5 or higher? Putting in some kind of warning about the limit on the default key seems sensible. Only place I can see putting it is in the settings help? Also in the welcome text?

I'm in two minds about making an option for the rate. Could limit it to 1-10 every second?

lordwelch commented on comictalker/talkers/comicvine.py:
```diff
@@ -150,27 +151,24 @@ class CVResult(TypedDict, Generic[T]):
     version: str
 
 
-CV_STATUS_RATELIMIT = 107
```
Is there a reason why this was removed? It should have been set to 429

mizaki replied:
It's no longer listed in the [API doc](https://comicvine.gamespot.com/api/documentation#toc-0-0) and as mentioned, they use the default HTTP code now.

lordwelch replied:
I meant the variable not the value

mizaki replied:
The var isn't used because it's now a HTTP status code like 200, 500, 404 etc.. That value was used after a CV had returned something. Now, it has moved from `_get_cv_content` to `_get_url_content`.

lordwelch replied:
Then please use `requests.status_codes.codes.TOO_MANY_REQUESTS`. Here is the rationale when this was originally added https://github.com/comictagger/comictagger/pulls?q=is%3Apr+is%3Aclosed+talker#discussion_r998881387
None of the numbers should be used be used but that is outside the scope of this PR, and 200, 500 and 404 are the most common and 404 is literally a meme at this point because everyone has used a dead link before. I don't think anyone that has ever used an http library has any doubt as to what those 3 mean; substantially less people know that 429 is the proper HTTP code for too many requests.

lordwelch commented on comictalker/talkers/comicvine.py:
```diff
@@ -151,31 +151,24 @@ class CVResult(TypedDict, Generic[T]):
-
-# https://comicvine.gamespot.com/forums/api-developers-2334/api-rate-limiting-1746419/
-# "Space out your requests so AT LEAST one second passes between each and you can make requests all day."
-limiter = Limiter(RequestRate(1, 2))
```
Lets make two limiters here one default limiter and a second custom limiter

lordwelch commented on comictalker/talkers/comicvine.py:
```diff
@@ -151,31 +151,24 @@ class CVResult(TypedDict, Generic[T]):
 
+    # https://comicvine.gamespot.com/forums/api-developers-2334/api-rate-limiting-1746419/
+    # "Space out your requests so AT LEAST one second passes between each and you can make requests all day."
+    limiter = Limiter(RequestRate(10, 10))
```
This should not be a class attribute move it to __init__ and assign the default limiter from the above comment

lordwelch commented on comictalker/talkers/comicvine.py:
```diff
@@ -217,8 +203,12 @@ def parse_settings(self, settings: dict[str, Any]) -> dict[str, Any]:
+
+        # Set a different limit if using the default API key
+        if self.api_key == self.default_api_key:
+            ComicVineTalker.limiter = Limiter(RequestRate(1, 2))
```
Here we want to cover both cases otherwise it will not change to the correct limiter every time

lordwelch commented on comictalker/talkers/comicvine.py:
```diff
@@ -436,41 +426,24 @@ def fetch_issues_by_series_issue_num_and_year(
     @limiter.ratelimit("cv", delay=True)
```
When using two different rate limiters we can't use the decorator format you will need to use the context manager https://github.com/vutran1710/PyrateLimiter#contextmanager

lordwelch reviewed (changes requested):
Lets set the default limit to (1, 5). Then add to the welcome text that the default ComicVine API Key is rate limited and getting a personal API Key will provide up to 5x speedup for online searches. That should provide sufficient motivation to get their own API Key. We also need to add a link to the api page https://comicvine.gamespot.com/api in the settings so that they know where to get one.

For now lets not make an option. If it proves to be useful in the future it will be simple enough to add later.

> which doesn't appear to be true. Which I interpreted to mean a gap of 1 second but 1,1 is probably correct.

How it works for (x,y) is that the first request marks the beginning of the time and then from that time you get x-1 more requests until y time has passed and then the next request that hasn't expired beyond y is used as the beginning so with (1,1) the first request goes through and starts the timer for 1 second 1-1=0 so you don't have any more requests until 1 second has passed. With (10,10) the first request goes through and starts the timer for 10 seconds 10-1=9 so there are 9 more requests that you can do in the next 10 seconds, if there are 10 more requests made at the 5 second mark, 9 go through without issue the 10th request waits for the remaining 5 seconds until the first request expires.

CV probably won't give any more specific info on it because they don't want people gaming the system

mizaki committed:
Create two module limiters and assign class limiter var depending. Add to welcome message limits of default CV API key.

mizaki commented:
I put a link in the welcome message to the wiki page. (I'm creating a new wiki page for Information Sources too [WIP](https://github.com/comictagger/comictagger/wiki/Comic-Information-Sources)). Because the settings are generated programmatically, I'm not sure on a way to get a pure label in there without making some bigger changes. Something could be added to the tooltip help but obviously isn't highly visible.

mizaki committed:
Use requests.status_codes.codes.TOO_MANY_REQUESTS

eduo commented:
Is there logging on the API at all? I'm on mac and checked both the built-in log window and the console but I can't find why I've got an error the last two days that the API access it not valid and no results.

Accessing the web itself works OK. I'd like to see what the API is responding, rather than comivine's interpretation.

I've been doing a ton of searches, refreshing my library's metadata, but for two days it hasn't worked any more.

If I try the API directly via URL I get proper results:

https://comicvine.gamespot.com/api/volumes/?api_key=XXXXX_APIKEY&format=json&sort=name:asc&filter=name:Walking%20Dead

<img width="552" alt="image" src="https://github.com/comictagger/comictagger/assets/154630/ca57bc37-b8d9-4ab3-9a55-9c5209a767b5">

lordwelch commented:
@eduo if you have a problem open an issue. Do not comment on old pull requests.
